DMTN-148.tex

```diff
@@ -50,9 +50,276 @@
+  \verb|CALIBDATE| & Creation & Concatenation of \verb|CALIB_CREATION_DATE| and \verb|CALIB_CREATION_TIME|. \\
+  \verb|VALIDSTART| & Certification & Start date the calibration is valid from.  \\
+  \verb|VALIDEND| & Certification & End date the calibration is valid from.  \\
+\end{tabular}
+
+
+\subsection{Calibration Construction in Gen3}
+
+Calibration products should be constructed using well defined
+PipelineTask code.  This removes the need to delineate between
+``calibrations'' and ``user-curated calibrations''.  The only
+remaining difference between these two groups is what the initial
+persisted data type is: ``calibrations'' are stored in FITS images, as
+they represent corrections that are only meaningful as full images;
+``user-curated calibrations'' can be visualized in a small number of
+parameters, and so persisting these in a text format to allow user
+inspection is reasonable.  ECSV and YAML are good formats for these
+calibrations, and a calibration's readText/writeText methods should
+implement one of these.
+
+In addition to the calibration product, the PipelineTask code must
+persist a record of the input data used for construction.  This, along
+with the already persisted pipeline definition and task configuration,
+allows the calibration to be regenerated by any user, making the
+construction process much more transparent than it currently is.  A
+YAML file containing the dataId parameters for all inputs would
+satisfy this.  Appending additional construction statistics (for
+example, weight factors applied to each input during the combination
+stage) would be helpful in the validation process.
+
+Newly generated calibration products will belong to a Gen3 collection,
+and that collection will be used to manage the calibration through the
+validation and certification process.  This collection may only
+contain one entry for a given OIDF set.
+
+\subsection{Validation}
+
+Calibration products are not currently validated, largely due to the
+opaque construction process.  Some set of proper validation tasks must
+be made, applying the newly constructed product to a set of test
+exposures prior to that calibration product's certification.  Applying
+these validation tasks continuously as new data is taken will inform
+the date ranges for which the product is certified.
+
+\subsection{Certification}
+
+Once a calibration product has been confirmed through validation to
+correct the instrument effect it is designed for, it will be certified
+with a date range denoting when the calibration is valid, and assigned
+to a particular calibration collection (which may contain calibrations
+the same OIDF values, as the validity range adds another dimension).
+For FITS image calibrations, this is equivalent to a database
+operation; no persisted data need change.  This is also acceptable for
+text calibrations, although conversion to FITS (using the
+CalibrationBaseClass methods) may be desired for IO speed. In
```

Comment:
The model for certification I am currently implementing in Gen3 (i.e. _not_ what's in Gen3 right now) does not make it possible to change the file format at certification time; certification will associate an existing dataset with a collection name and a validity range, but not write a new dataset.

I think that's inconsistent with the proposal to initially write some calibration datasets in text formats, unless those text formats are also fast enough to be used later.  Given that butler data repositories are already generally considered to be opaque and accessible only through butler clients, my preferred resolution to this would be to always write datasets in the faster, binary (e.g. usually FITS) format, and consider text formats an export format.

An alternative would be to define `PipelineTasks` that read one dataset type (in text format) and write an equivalent one (with a different name, in binary format); these tasks would be run just before the collection-level certification operation.  That would maintain the provenance chain through the format change, but the extra names, extra tasks, and extra steps are clear downsides.

Yet another alternative would be to have different Datastores for text-format and binary-format datasets; we could set these up to write to both initially, and read first from the binary-format one (and perhaps consider the text-format ones temporary, disposing of them entirely with some cadence).  That seems much better than the second option, but it's still a lot of setup and teardown code we'd end up adding to the high-level calibration production tooling relative to the first.

All of these options can coexist, of course, provided both formats have butler `Formatter` implementations, but if we can get away with only approach, all the better.

Reply:
I've come around to agree with this: we always write FITS as part of the calibration production, although support text formats for export.  I think this means that calibrations that are produced and certified entirely within one butler only have valid date ranges defined in that butler, and that the export process writes those valid ranges to the text format.

Reply:
I still don't feel right about writing the validity ranges to the text files. We deliberately designed obs_*_data so that we didn't need to edit calibrations just to change validity.